packages/protocol/contracts/L1/verifiers/SgxVerifier2.sol

```diff
@@ -0,0 +1,139 @@
+
+    mapping(address instance => uint256 registeredAt) public instances;
+
+    uint256[49] private __gap;
+
+    event InstanceRegistered(
+        address indexed replaced, address indexed instance, uint256 registeredAt
+    );
+
+    error SGX_INVALID_AUTH();
+    error SGX_INVALID_INSTANCE();
+    error SGX_INVALID_PROOF_SIZE();
+    error SGX_INSTANCE_REGISTERED();
+
+    /// @notice Initializes the contract with the provided address manager.
+    /// @param _addressManager The address of the address manager contract.
+    function init(address _addressManager) external initializer {
+        EssentialContract._init(_addressManager);
+    }
+
+    function registerInstance(address instance) external onlyOwner {
+        _replaceInstance(address(0), instance);
+    }
+
+    // TODO(dani): who will call this function?
+    function replaceInstance(
+        address newInstance,
+        bytes memory signature
+    )
+        external
+    {
+        address oldInstance = keccak256(
+            abi.encode("REGISTER_SGX_INSTANCE", newInstance)
+        ).recover(signature);
+
+        _replaceInstance(oldInstance, newInstance);
+    }
+
+    /// @inheritdoc IVerifier
+    function verifyProof(
+        uint64,
+        address prover,
+        bool isContesting,
+        TaikoData.BlockEvidence calldata evidence
+    )
+        external
+        onlyFromNamed("taiko")
+    {
+        // Do not run proof verification to contest an existing proof
+        if (isContesting) return;
+
+        // Size is: 87 bytes
+        // 2 bytes + 20 bytes + 65 bytes = 87
+        if (evidence.proof.length <= 20) {
+            revert SGX_INVALID_PROOF_SIZE();
+        }
+
+        address newInstance =
+            address(bytes20(LibBytesUtils.slice(evidence.proof, 0, 20)));
+
+        bytes memory signature = LibBytesUtils.slice(evidence.proof, 20);
+
+        address oldInstance =
+            getSignedHash(evidence, prover, newInstance).recover(signature);
+
+        _replaceInstance(oldInstance, newInstance);
+    }
+
+    function isInstanceValid(address instance) public view returns (bool) {
+        return instances[instance] + INSTANCE_EXPIRY > block.timestamp;
+    }
+
+    function getSignedHash(
+        TaikoData.BlockEvidence memory evidence,
+        address prover,
+        address newAddress
+    )
+        public
+        pure
+        returns (bytes32 signedHash)
+    {
+        return keccak256(
+            abi.encode(
+                evidence.metaHash,
+                evidence.parentHash,
+                evidence.blockHash,
+                evidence.signalRoot,
+                evidence.graffiti,
+                prover,
+                newAddress
+            )
+        );
+    }
+
+    function _replaceInstance(
+        address oldInstance,
+        address newInstance
+    )
+        private
+    {
+        if (oldInstance != address(0)) {
+            if (!isInstanceValid(oldInstance)) revert SGX_INVALID_AUTH();
```

Comment:
The SGX instance shall have a chance to update/replace his own 'old' instance even in case it is expired.